server/middlewares/rateLimiter.ts

```diff
@@ -0,0 +1,42 @@
+import { Context, Next } from "koa";
+import { RateLimiterRedis } from "rate-limiter-flexible";
+import env from "@server/env";
+import { RateLimitExceededError } from "@server/errors";
+import Redis from "@server/redis";
+
+export default function rateLimiter() {
+  // TODO: Adjust as per actual throughput data later
+  const MAX_ALLOWED_REQUESTS = 5000;
+  const DURATION_WINDOW = 1; // in seconds
+
+  const opts = {
+    storeClient: Redis.defaultClient,
+    keyPrefix: "rateLimiter",
+    points: MAX_ALLOWED_REQUESTS,
+    duration: DURATION_WINDOW,
+  };
+
+  return async function rateLimiterMiddleware(ctx: Context, next: Next) {
+    if (!env.ENABLE_RATE_LIMITER) {
+      return next();
+    }
+
+    const limiter = new RateLimiterRedis(opts);
+
+    try {
+      await limiter.consume(ctx.ip);
+    } catch (rateLimiterRes) {
+      ctx.set("Retry-After", `${rateLimiterRes.msBeforeNext / 1000}`);
+      ctx.set("X-RateLimit-Limit", `${opts.points}`);
+      ctx.set("X-RateLimit-Remaining", `${rateLimiterRes.remainingPoints}`);
+      ctx.set(
+        "X-RateLimit-Reset",
```

Comment:
```suggestion
      ctx.set("RateLimit-Limit", `${opts.points}`);
      ctx.set("RateLimit-Remaining", `${rateLimiterRes.remainingPoints}`);
      ctx.set(
        "RateLimit-Reset",
```

`X` prefix no longer recommended on headers

Reply:
Hmm...Didn't know that. Will patch it up.

Reply:
https://datatracker.ietf.org/doc/html/rfc6648

Reply:
Fixed!